### PR TITLE
ethereum: allow mainnet on testnet keypath and vice versa

### PR DESCRIPTION
**Pull request by benma**
Until now:

Mainnet coins were under the xpub: m/44'/60'/0'/0
Testnet coins were under the xpub: m/44'/1'/0'/0

With this commit, one can use either path with any Ethereum
network. If it is not matching the above, a warning is displayed.

The reason this is changed is that some wallets, like Metamask, use
the 60' path for all networks. Using testnet coins on Metamask hence
fails since the BitBox02 rejected the transactions based on the
keypath.

Mixing keypaths like this is safe even if the user ignores the
warning, as the chain id (unique per ETH network) is part of the
transaction sighash, so one cannot be tricked into sending mainnet
coins instead of testnet coins. The only risk is that mainnet coins
might not be found if sent to the testnet keypath, hence the warning.

### CODE REVIEW

**Review by x1ddos — APPROVED** [review]
LGTM

File: src/rust/bitbox02-rust/src/hww/api/ethereum.rs [hunk 1]
```diff
@@ -18,6 +18,7 @@ compile_error!(
 );
 
 mod amount;
+mod common;
```
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02-rust/src/hww/api/ethereum.rs:21]: I want to merge bitbox02_rust::apps::bitcoin with bitbox02_rust::hww::api::bitcoin, and merge the toplevel apps crate (that only contains a few eth funcs) to bitbox02_rust::hww::api::ethereum. I don't see the need to have multiple btc/eth crates/modules.

File: src/rust/bitbox02-rust/src/hww/api/ethereum/common.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+/// If the second element of `keypath` does not match the expected bip44 coin value for the given
+/// coin, we warn the user about an unusual keypath.
+///
+/// The keypath is already assumed to be validated and that the second element is one of the
+/// Ethereum bip44 values, e.g. `60'` or `1'`.
+///
+/// A warning suffices so the user does not accidentally send e.g. mainnet coins to a testnet path
+/// (m/44'/1'/...). It is safe to make a transaction on the 'wrong' keypath as the chain id is
+/// unique and part of the transaction sighash.
+pub async fn maybe_warn_keypath(
```
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02-rust/src/hww/api/ethereum/common.rs:37]: With "maybe" I wanted to make clear that there might or might not be a warning. With `warn_...` it sounds like it was already determined that there should be a warning and this function would show it.

I'll rename it to `warn_unusual_keypath` though

File: src/rust/bitbox02-rust/src/hww/api/ethereum/common.rs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    params: &Params,
+    title: &str,
```
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02-rust/src/hww/api/ethereum/common.rs:39]: It isn't in the case of ERC20 tokens. I will improve this anyway, but first https://github.com/digitalbitbox/bitbox02-firmware/pull/781, so the name is with the rest of the coin params and does not have special treatment.

File: src/rust/bitbox02-rust/src/hww/api/ethereum/common.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02-rust/src/hww/api/ethereum/common.rs]: Done

**Review by x1ddos — APPROVED** [review]

File: src/rust/bitbox02-rust/src/hww/api/ethereum.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02-rust/src/hww/api/ethereum.rs]: common.rs is now keypath.rs - the other keypath functions will eventually be moved here as well

**Pull request merged**